### ISSUE TITLE
Mappers should reject keys

### ISSUE THREAD

**Issue by gotar** [rom-rb/rom#194]
```ruby
require 'rom'

ROM.setup(:memory)

class Users < ROM::Relation[:memory]
end

class UserMapper < ROM::Mapper
  relation :users
  register_as :entity

  attribute :name
  attribute :age
end

rom = ROM.finalize.env

rom.relations.users << { name: 'Oskar', age: 31, phone: '123456' }

user = rom.relation(:users).map_with(:entity).first

user == { name: 'Oskar', age: 31 } # => false but should be true
```

**Comment by cflipse** [comment]
If there are no attributes specified, what's the expected behavior?

And how does this mesh with the `exclude` option?

**Comment by gotar** [comment]
exclude change nothing (not work)

```ruby
require 'rom'

ROM.setup(:memory)

class Users < ROM::Relation[:memory]
end

class UserMapper < ROM::Mapper
  relation :users
  register_as :entity

  attribute :name
  attribute :age
  exclude :phone
end

rom = ROM.finalize.env

rom.relations.users << { name: 'Oskar', age: 31, phone: '123456' }

user = rom.relation(:users).map_with(:entity).first

p user #=> {:name=>"Oskar", :age=>31, :phone=>"123456"}
```

**Comment by gotar** [comment]
the same without attributes at all

```
require 'rom'

ROM.setup(:memory)

class Users < ROM::Relation[:memory]
end

class UserMapper < ROM::Mapper
  relation :users
  register_as :entity
end

rom = ROM.finalize.env

rom.relations.users << { name: 'Oskar', age: 31, phone: '123456' }

user = rom.relation(:users).map_with(:entity).first

p user #=> {:name=>"Oskar", :age=>31, :phone=>"123456"}
```

**Comment by gotar** [comment]
As I know outside model (without model) `attribues` works only to transform, do not drop anything

**Comment by solnic** [comment]
`exclude` is needed when you inherit from another mapper and want to exclude some inherited attributes. But overall mappers do not reject keys, `attribute` dsl is only used for transformations and coercions. We need to figure out how to improve that, I can see that rejecting could be handy.

**Comment by gotar** [comment]
yeap right now when we, from DB after create, got everything will be nice to have tool to easly map it to hash we like to have

**Comment by gotar** [comment]
Generally mapper for me should works like Virtus, if you don't specify any attribute then it return `{}`. You need to specify whole complex structure you want to have after transformation.

**Comment by solnic** [comment]
The thing is that a mapper becomes almost a no-op if there are no attributes specified, which makes it fast. I'd say we should have a way to tell the mapper to actually reject undefined attributes.

A typical scenario is that a relation returns tuples that are already in the expected state and doing anything extra with that is just an overhead that you don't need. My point is that rejecting attributes could be considered as a special case. For instance if you don't want certain keys to be in a tuple then in many many cases your db supports that (ie projection in RDBMS through `select foo, bar ... `). Only if a given data backend is dumb and returns more data that you need *and for some reason* it's a problem for you then, and only then, you will need a key rejection mechanism in a mapper.

FWIW I had this case only once when mapping to anima objects because anima raises an error when there are unexpected keys in a tuple.

Please remember that mapper layer should be used when your db is lacking functionality wrt data projections.

**Comment by gotar** [comment]
those `select ...` works ok for fetching data from DB, but problem I have is with data returned after `create command`. We don't have any control on db side what fields we will get, and then need mapper to fix that.

**Comment by gotar** [comment]
But maybe we try to fix it from wrong side and we should return nothing from create call (as CQRS said) then we will have normal db read call there or allow to inject call commands do after create

**Comment by solnic** [comment]
Welllllllll you can always do this:

``` ruby
require 'rom'

ROM.setup(:sql, ['sqlite::memory'])

ROM.boot.repositories[:default].connection.create_table(:users) do
  primary_key :id
  String :name
end

class Users < ROM::Relation[:sql]
  def by_id(id)
    where(id: id)
  end

  def from_created(users)
    by_id(users.map { |user| user[:id] })
  end
end

class CreateUser < ROM::Commands::Create[:sql]
  relation :users
  register_as :create
end

class UserMapper < ROM::Mapper
  relation :users
  register_as :entity
  model name: 'User'
  attribute :id
  attribute :name
end

rom = ROM.finalize.env

command = rom.command(:users).create.with(id: 1, name: 'Jane')
command >>= rom.relation(:users).from_created
command >>= rom.mappers[:users].entity

puts command.call.first.inspect
#<User:0x007faaa36d0308 @id=1, @name="Jane">
```

:smiley:

**Comment by gotar** [comment]
:mindblow: again I realize I'm still beginner in ROM universe
